hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/ethereum/CompositeEthereumTransactionParser.java

```
@@ -0,0 +1,57 @@
+ * ​
+ * Copyright (C) 2019 - 2022 Hedera Hashgraph, LLC
+ * ​
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.esaulpaugh.headlong.rlp.RLPDecoder;
+import javax.inject.Named;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.context.annotation.Primary;
+
+import com.hedera.mirror.common.domain.transaction.EthereumTransaction;
+
+@Log4j2
+@Named
+@Primary
+@RequiredArgsConstructor
+public class CompositeEthereumTransactionParser implements EthereumTransactionParser {
+    private final LegacyEthereumTransactionParser legacyEthereumTransactionParser;
+    private final Eip1559EthereumTransactionParser eip1559EthereumTransactionParser;
+
+    @Override
+    public EthereumTransaction parse(byte[] transactionBytes) {
+        var ethereumTransactionParser = getEthereumTransactionParser(transactionBytes);
+        return ethereumTransactionParser.parse(transactionBytes);
+    }
+
+    @Override
+    public byte[] retrievePublicKey(EthereumTransaction ethereumTransaction) {
+        return ethereumTransaction.getType() == 0 ?
+                legacyEthereumTransactionParser.retrievePublicKey(ethereumTransaction) :
+                eip1559EthereumTransactionParser.retrievePublicKey(ethereumTransaction);
+    }
+
+    private EthereumTransactionParser getEthereumTransactionParser(byte[] transactionBytes) {
+        var decoder = RLPDecoder.RLP_STRICT.sequenceIterator(transactionBytes);
```

Does this have to decode the entire bytes before it returns or is it lazy and only decodes as needed?

---

I believe entire bytes.
Looking for a more performant way to determine which type so we only decode once.
However, it breaks the interface to use the result of the first decode.

---

Another options is to try `Eip1559` first and only call Legacy on failure.
That way it's either a single parse of bytes of a fail fast decode attempt followed by a full bytes parse.

---

I mean, it's just a byte array. You can use sample bytes to see how legacy vs EIP1559 looks then just check manually for those byte difference at the beginning.

---

Simplified